### PR TITLE
sysconfig/cloudinit: add RestrictCloudInit

### PR DESCRIPTION
**Pull request by mvo5**
RestrictCloudInit will implement the cloud-init specific bits of mitigation against CVE-2020-11933, insofar as it will disable importing of cloud-init NoCloud datasources from arbitrary filesystems that an attacker could control and put malicious cloud-init data on as well as always limit the datasource for cloud-init to use on subsequent boots to the detected datasource.

We have to parse the cloud-init status.json file in order to determine what datasource was used because while the cloud-init status does give us the same information, it is in a less useful format that is harder to parse than the JSON here, and the JSON here is in a stable v1 structure that can be relied upon.

This is #3 from the snapd-private repo used to address the cloud-init fix.

Build on top of  #9023

### CODE REVIEW

File: sysconfig/cloudinit.go [hunk 1]
```diff
@@ -88,3 +92,220 @@ func configureCloudInit(opts *Options) (err error) {
...
+		// these states are all
+		return CloudInitEnabled, nil
+	}
+}
+
+type v1Data struct {
+	Datasource string
+}
+
+type cloudInitStatus struct {
```
Comment (zyga) [inline sysconfig/cloudinit.go:192]: This structure warrants a comment that it is externally defined. I would also add JSON struct tags even though they are redundant, to highlight this.

File: sysconfig/cloudinit.go [hunk 2]
```diff
@@ -88,3 +92,220 @@ func configureCloudInit(opts *Options) (err error) {
...
+	// "running" and "not run" are considered Enabled, see doc-comment
+	case "running", "not run":
+		fallthrough
+	default:
+		// these states are all
+		return CloudInitEnabled, nil
+	}
+}
+
+type v1Data struct {
```
Comment (zyga) [inline sysconfig/cloudinit.go:188]: This structure warrants a comment that it is externally defined. I would also add JSON struct tags even though they are redundant, to highlight this.

File: sysconfig/cloudinit_test.go [hunk 3]
```diff
@@ -80,3 +81,321 @@ func (s *sysconfigSuite) TestCloudInitInstalls(c *C) {
...
+
+	for _, t := range tt {
+		// setup status.json
+		old := dirs.GlobalRootDir
+		dirs.SetRootDir(c.MkDir())
+		defer func() { dirs.SetRootDir(old) }()
+		statusJSONFile := filepath.Join(dirs.GlobalRootDir, "/run/cloud-init/status.json")
+		if t.cloudInitStatusJSON != "" {
+			err := os.MkdirAll(filepath.Dir(statusJSONFile), 0755)
+			c.Assert(err, IsNil, Commentf(t.comment))
```
Comment (zyga) [inline sysconfig/cloudinit_test.go:355]: Nitpick, `Commentf(t.comment)` could be computed once per loop.

File: sysconfig/cloudinit.go [hunk 4]
```diff
@@ -88,3 +92,220 @@ func configureCloudInit(opts *Options) (err error) {
...
+
+	switch res.Datasource {
+	case "NoCloud":
+		// With the NoCloud datasource, we also need to restrict/disable the
+		// import of arbitrary filesystem labels to use as datasources, i.e. a
+		// USB drive inserted by an attacker with label CIDATA will defeat
+		// security measures on Ubuntu Core, so with the additional fs_label
+		// spec, we disable that import.
+		err := ioutil.WriteFile(cloudInitRestrictFile, nocloudRestrictYaml, 0644)
+		if err != nil {
```
Comment (zyga) [inline sysconfig/cloudinit.go:296]: Nitpick, the `if err != nil` ... part here and below could be removed if the assignment had reused `err` and simply returned `res, err` at the end of the function.

**Comment by anonymouse64** [comment]
I merged master here so the diff is both smaller since a pre-req for this PR was already merged master, and so we can eliminate the seccomp failures. There are now expected groovy failures for pulseaudio/audio-{playback,record} and the preseed PR's, but beyond that this should be mergable. If everything is green except groovy I will merge this in my PM

**Pull request merged**